### DISCUSSION TITLE
instantiate the proper object (Eth / WiFi) at run time and not at compilation time

### DISCUSSION THREAD

**Discussion by infrafast** [skaarj1989/mWebSockets#67] [Ideas]
This would cover the need of having a controller with both ethernet and Wifi connectivity without having to instantiate multiple object and double the code

**Comment by skaarj1989** [comment]
```cpp

IWebSocketServer *server{nullptr};

void setup() {
  if (...) {
    server = new WebSocketServer<EthernetServer, EthernetClient, 4>;
  }
  else {
    server = new WebSocketServer<WiFiServer, WiFiClient, 4>;
  }
}

void loop() {
  server->listen();
}
```

Do you expect something like that?

**Reply by infrafast** [reply]
yes that would cover the need.

plus also implementation of the callback function :
**_ wss.onConnection([](MyWebSocket &ws) {
...
    ws.onMessage([](MyWebSocket &ws, const WebSocketDataType dataType,
                    const char *message, uint16_t length) {_**

In this case I imagine we would also need to use 

WebSocket*  socket{nullptr} instead of MyWebSocket and then provide an instance of the websocket to be either

socket = new WebSocket<EthernetClient>

or 

**_socket = new WebSocket<WiFiClient>_**

------------------------------------------ ALTERNATIVE maybe more comprehensive would be to use it without template
server = new WebSocketServer(EthernetServer, EthernetClient, 4);

**Comment by skaarj1989** [comment]
Fixed. Try the following code:

```cpp
#include <WiFi.h> // or ESP8266WiFi.h

using WiFiWebSocketServer = WebSocketServer<WiFiServer, WiFiClient, 4>;

#include <Ethernet.h>

using EthernetWebSocketServer = WebSocketServer<EthernetServer, EthernetClient, 4>;

IWebSocketServer *wss{ nullptr };

void setup() {
  // setup network ...

  if (useEthernet) {
    wss = new EthernetWebSocketServer{};
  } else {
    wss = new WiFiWebSocketServer{};
  }

  wss->onConnection([](IWebSocket &ws) {
    const auto protocol = ws.getProtocol();
    if (protocol) {
      _SERIAL.print(F("Client protocol: "));
      _SERIAL.println(protocol);
    }

    ws.onMessage([](IWebSocket &ws, const WebSocketDataType dataType,
                    const char *message, uint16_t length) {
      switch (dataType) {
        case net::WebSocketDataType::TEXT:
          _SERIAL.print(F("Received: "));
          _SERIAL.println(message);
          break;
        case net::WebSocketDataType::BINARY:
          _SERIAL.println(F("Received binary data"));
          break;
      }

      ws.send(dataType, message, length);
    });

    ws.onClose(
      [](IWebSocket &, const WebSocketCloseCode, const char *, uint16_t) {
        _SERIAL.println(F("Disconnected"));
      });

    _SERIAL.print(F("New client: "));
    _SERIAL.println(ws.getRemoteIP());

    const char message[]{ "Hello from Arduino server!" };
    ws.send(WebSocketDataType::TEXT, message, strlen(message));
  });
  wss->begin();
}

void loop() {
  wss->listen();
}
```

**Reply by infrafast** [reply]
Great thanks for the adapation,  I was sick and could not test earlier so apologize for my late feedback. 

I can confirm this compiles fine and works for wifi like a charm.  The current test envirnment I have do not allow me to test ETH yet but I will make deeper test this week end and reports here.

**Reply by infrafast** [reply]
test done on a W5500 chip, works like a charm :-) thanks!

**Reply by infrafast** [reply]
Hello I am using the lib since a while now and would have the exact same need to instantiate client objet the same way. WOuld you be ok to add IWebSocketClient with callbacks ? 
Thanks a lot

**Comment by infrafast** [comment]
Hi, it seems that calling shutdown on wss is crashing:

wss->shutdown();

log:

CORRUPT HEAP: Bad head at 0x3fca6f50. Expected 0xabba1234 got 0x3fca3784
assert failed: multi_heap_free multi_heap_poisoning.c:259 (head != NULL)

while when we call delete wss , which I do believe calls shutdown in turn, it executes properly.
Maybe something linked to templates/virtual method call?

Is it something I should further investigate or maybe you have a clue ?
Thank you
BR
Thomas

**Reply by skaarj1989** [reply]
Here's the issue (I think):
https://github.com/skaarj1989/mWebSockets/blob/321b2898b5a2aa016042ba60308fbc7706d0d1cb/src/WebSocketServer.inl#L31

You can easily fix it with:
```cpp
 for (auto &ws : m_sockets)
```

**Reply by infrafast** [reply]
Hllo, thanks for the fix, it doesnt crash anymore upon call of shutdown(). 
This however doesnt not fix the root problem that I tought it would fix. I am opening a new comment below to avoid mixing things.

**Comment by infrafast** [comment]
Hi, again me. I have a function setWSServerwhich is triggered upon network connection change (either Wifi or Ethernet connect, so connection would be tggled. When Eth is on, Wifi is off and vice versa)

```c++
 IWebSocketServer *wss{ nullptr };   

    void setWSServer(){

      if (wss != nullptr){
        wss->shutdown();
        delete wss; 
      }
      
      if (hwm->configMedia->media == CABLE_MEDIA) wss = new EthernetWebSocketServer{};
      else wss = new WiFiWebSocketServer{};

      wss->onConnection(socketConnectionCallback);    
      wss->begin();

    }

```

It works ok, when you disconnect Ethernet, Wifi take over and when you unplug Eth cable, Wifi Take over and sockets are working fine.
However, it takes >5s to recover the wifi connection, looks like there is a timeout somewhere. 


Side observation :
1) the "Server" used by WSS (either Eth or Wifi) is created in another portion of the code but begin() and eth would prevents socket to work, see below
```C++
if (server == NULL) server = new EthernetServer(80);
if (media != CABLE_MEDIA) server->begin(); // <-- for some reason this begin cause the webserver not react with ethernet if it is called. But is required for WiFi
```

2) I suspect the wss desctructor to trigger a kind of timeout as 
 a)if I remove the wss.begin() this is still "blocking" for a while and then code continue (having of course the socket not being able to reconnect)
 b) when I moved the       if (wss != nullptr){
        wss->shutdown();
        delete wss; 
      }
 in another place in my code this will block at that place
3) the time out doesnt not comes from the PHY connection as such, I cheked this is ok

Your support is as usual much appreciated
